### PR TITLE
implement atomic swap 100 close to specification

### PR DESCRIPTION
**Pull request by EmilGeorgiev**

### CODE REVIEW

File: modules/apps/100-atomic-swap/keeper/msg_server_make_swap.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		return nil, err0
+	}
+
+	sender, err1 := sdk.AccAddressFromBech32(msg.MakerAddress)
+	if err1 != nil {
+		return nil, err1
+	}
+
+	balance := k.bankKeeper.GetBalance(ctx, sender, msg.SellToken.Denom)
+	if balance.Amount.BigInt().Cmp(msg.SellToken.Amount.BigInt()) < 0 {
```
Comment (liangping) [inline modules/apps/100-atomic-swap/keeper/msg_server_make_swap.go]: Amount has `Gt()` an `Lt()` functions. Would be better to change

File: modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"context"
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/gogo/protobuf/proto"
+	"github.com/ibcswap/ibcswap/v6/modules/apps/100-atomic-swap/types"
+)
+
+// TakeSwap is the step 5 (Lock Order & Lock Token) of the atomic swap: https://github.com/liangping/ibc/blob/atomic-swap/spec/app/ics-100-atomic-swap/ibcswap.png
```
Comment (liangping) [inline modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go:12]: Would be better to use URL of cosmos repo. Https://github.com/cosmos/ibc

File: modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	if order.Status != types.Status_SYNC && order.Status != types.Status_INITIAL {
+		return nil, errors.New("order is not in valid state")
+	}
+
+	// Make sure the maker's buy token matches the taker's sell token
+	if order.Maker.BuyToken.Denom != msg.SellToken.Denom || !order.Maker.BuyToken.Amount.Equal(msg.SellToken.Amount) {
+		return &types.MsgTakeSwapResponse{}, errors.New("invalid sell token")
+	}
+
+	// Checks if the order has already been taken
```
Comment (liangping) [inline modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go:45]: I am thinking to add a new LOCK status for order. And also add status changing here.

File: modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		return &types.MsgTakeSwapResponse{}, errors.New("invalid taker address")
+	}
+
+	takerAddr, err := sdk.AccAddressFromBech32(msg.TakerAddress)
+	if err != nil {
+		return &types.MsgTakeSwapResponse{}, err
+	}
+
+	balance := k.bankKeeper.GetBalance(ctx, takerAddr, msg.SellToken.Denom)
+	if balance.Amount.BigInt().Cmp(msg.SellToken.Amount.BigInt()) < 0 {
```
Comment (liangping) [inline modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go]: The same issue

File: proto/ibc/applications/atomic_swap/v1/tx.proto [hunk 5]
```diff
@@ -10,12 +10,12 @@ import "ibc/core/client/v1/client.proto";
 
 // Msg defines the ibc/swap Msg service.
 service Msg {
-  rpc MakeSwap(MsgMakeSwapRequest) returns (MsgMakeSwapResponse);
-  rpc TakeSwap(MsgTakeSwapRequest) returns (MsgTakeSwapResponse);
-  rpc CancelSwap(MsgCancelSwapRequest) returns (MsgCancelSwapResponse);
+  rpc MakeSwap(MakeSwapMsg) returns (MsgMakeSwapResponse);
```
Comment (liangping) [inline proto/ibc/applications/atomic_swap/v1/tx.proto:13]: Seems that Use prefix Msg is more common than suffix in cosmos. 

I know that's the problem of spec.

**Review by liangping — CHANGES_REQUESTED** [review]
It's much better.  Few suggestion to be improved

File: modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go (hunk 3 above)
Comment (EmilGeorgiev) [reply modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go:45]: I can make this in separate PR, this is already too big

File: modules/apps/100-atomic-swap/keeper/msg_server_make_swap.go (hunk 1 above)
Comment (EmilGeorgiev) [reply modules/apps/100-atomic-swap/keeper/msg_server_make_swap.go]: Done

File: modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go (hunk 2 above)
Comment (EmilGeorgiev) [reply modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go:12]: Done

File: modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go (hunk 4 above)
Comment (EmilGeorgiev) [reply modules/apps/100-atomic-swap/keeper/msg_server_take_swap.go]: Done

File: proto/ibc/applications/atomic_swap/v1/tx.proto (hunk 5 above)
Comment (EmilGeorgiev) [reply proto/ibc/applications/atomic_swap/v1/tx.proto:13]: Can I make this in a separate issue and separate PR? I also can update the specification

Comment (liangping) [reply proto/ibc/applications/atomic_swap/v1/tx.proto:13]: Yeah. That would be better.

**Pull request merged**